File: services/audit-service/tests/AuditService.Tests/JsonSchemaAssert.cs
using System.Globalization;
using System.Text.Json;

namespace AuditService.Tests;

/// <summary>
/// Minimal JSON Schema (draft-07 subset) checker for contract tests: type, const, enum,
/// required, properties, additionalProperties, items, minItems, local $ref and the
/// uuid/date-time formats.
/// </summary>
public sealed class JsonSchemaAssert
{
    private readonly JsonElement _root;

    private JsonSchemaAssert(JsonElement root) => _root = root;

    public static JsonSchemaAssert Load(string schemaFile)
    {
        using var doc = JsonDocument.Parse(File.ReadAllText(SharedSchemaPath(schemaFile)));
        return new JsonSchemaAssert(doc.RootElement.Clone());
    }

    public JsonElement Definition(string name)
    {
        if (!_root.GetProperty("definitions").TryGetProperty(name, out var def))
        {
            throw new Xunit.Sdk.XunitException($"Definition '{name}' not found in schema");
        }
        return def;
    }

    public void ConformsTo(JsonElement schema, JsonElement instance)
    {
        var violations = Validate(schema, instance, "$");
        if (violations.Count > 0)
        {
            throw new Xunit.Sdk.XunitException(
                $"Payload violates contract:\n  {string.Join("\n  ", violations)}\nPayload: {instance}");
        }
    }

    public List<string> Validate(JsonElement schema, JsonElement instance, string path)
    {
        var violations = new List<string>();

        if (schema.TryGetProperty("$ref", out var reference))
            violations.AddRange(Validate(Resolve(reference.GetString()!), instance, path));

        if (schema.TryGetProperty("const", out var constant) && !JsonEquals(constant, instance))
            violations.Add($"{path}: expected const {constant}, got {instance}");

        if (schema.TryGetProperty("enum", out var allowedValues) &&
            !allowedValues.EnumerateArray().Any(v => JsonEquals(v, instance)))
            violations.Add($"{path}: {instance} not in enum {allowedValues}");

        if (schema.TryGetProperty("type", out var typeNode))
        {
            var allowed = typeNode.ValueKind == JsonValueKind.Array
                ? typeNode.EnumerateArray().Select(t => t.GetString()!).ToList()
                : new List<string> { typeNode.GetString()! };
            if (!allowed.Any(t => MatchesType(t, instance)))
                violations.Add($"{path}: expected type [{string.Join(", ", allowed)}], got {instance.ValueKind}");
        }

        if (instance.ValueKind == JsonValueKind.String && schema.TryGetProperty("format", out var format))
        {
            var text = instance.GetString()!;
            switch (format.GetString())
            {
                case "uuid" when !Guid.TryParse(text, out _):
                    violations.Add($"{path}: '{text}' is not a uuid");
                    break;
                case "date-time" when !DateTimeOffset.TryParseExact(
                    text,
                    new[] { "yyyy-MM-dd'T'HH:mm:ssK", "yyyy-MM-dd'T'HH:mm:ss.FFFFFFFK" },
                    CultureInfo.InvariantCulture,
                    DateTimeStyles.None,
                    out _):
                    violations.Add($"{path}: '{text}' is not an RFC 3339 date-time");
                    break;
            }
        }

        if (instance.ValueKind == JsonValueKind.Object)
        {
            var hasProperties = schema.TryGetProperty("properties", out var properties);
            if (schema.TryGetProperty("required", out var required))
            {
                foreach (var name in required.EnumerateArray().Select(r => r.GetString()!))
                {
                    if (!instance.TryGetProperty(name, out _))
                        violations.Add($"{path}: missing required property '{name}'");
                }
            }

            var hasAdditional = schema.TryGetProperty("additionalProperties", out var additional);
            foreach (var property in instance.EnumerateObject())
            {
                if (hasProperties && properties.TryGetProperty(property.Name, out var propertySchema))
                    violations.AddRange(Validate(propertySchema, property.Value, $"{path}.{property.Name}"));
                else if (hasAdditional && additional.ValueKind == JsonValueKind.Object)
                    violations.AddRange(Validate(additional, property.Value, $"{path}.{property.Name}"));
                else if (hasAdditional && additional.ValueKind == JsonValueKind.False)
                    violations.Add($"{path}: property '{property.Name}' is not declared in the contract");
            }
        }

        if (instance.ValueKind == JsonValueKind.Array)
        {
            var length = instance.GetArrayLength();
            if (schema.TryGetProperty("minItems", out var minItems) && length < minItems.GetInt32())
                violations.Add($"{path}: expected at least {minItems.GetInt32()} items, got {length}");

            if (schema.TryGetProperty("items", out var itemSchema))
            {
                var i = 0;
                foreach (var item in instance.EnumerateArray())
                    violations.AddRange(Validate(itemSchema, item, $"{path}[{i++}]"));
            }
        }

        return violations;
    }

    private JsonElement Resolve(string pointer)
    {
        if (!pointer.StartsWith("#/", StringComparison.Ordinal))
            throw new Xunit.Sdk.XunitException($"Only local $ref pointers are supported, got '{pointer}'");

        var node = _root;
        foreach (var segment in pointer[2..].Split('/'))
        {
            var key = segment.Replace("~1", "/").Replace("~0", "~");
            if (!node.TryGetProperty(key, out node))
                throw new Xunit.Sdk.XunitException($"$ref '{pointer}' does not resolve");
        }
        return node;
    }

    private static bool MatchesType(string type, JsonElement instance) => type switch
    {
        "object" => instance.ValueKind == JsonValueKind.Object,
        "array" => instance.ValueKind == JsonValueKind.Array,
        "null" => instance.ValueKind == JsonValueKind.Null,
        "string" => instance.ValueKind == JsonValueKind.String,
        "boolean" => instance.ValueKind is JsonValueKind.True or JsonValueKind.False,
        "integer" => instance.ValueKind == JsonValueKind.Number && instance.TryGetInt64(out _),
        "number" => instance.ValueKind == JsonValueKind.Number,
        _ => false,
    };

    private static bool JsonEquals(JsonElement a, JsonElement b) => a.GetRawText() == b.GetRawText();

    private static string SharedSchemaPath(string schemaFile)
    {
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir is not null)
        {
            var candidate = Path.Combine(dir.FullName, "shared", "events", "schemas", schemaFile);
            if (File.Exists(candidate)) return candidate;
            dir = dir.Parent;
        }
        throw new Xunit.Sdk.XunitException(
            $"Could not locate shared/events/schemas/{schemaFile} above {AppContext.BaseDirectory}");
    }
}

File: services/notification-service/src/test/kotlin/com/otterworks/notification/contract/JsonSchemaAssert.kt
package com.otterworks.notification.contract

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.time.OffsetDateTime
import java.util.UUID
import kotlin.test.fail

/**
 * Minimal JSON Schema (draft-07 subset) checker for contract tests: type, const, enum,
 * required, properties, additionalProperties, items, minItems and the uuid/date-time formats.
 */
object JsonSchemaAssert {

    fun loadDefinition(schemaFile: String, definition: String): JsonObject {
        val schema = Json.parseToJsonElement(Files.readString(sharedSchemaPath(schemaFile))).jsonObject
        return schema["definitions"]?.jsonObject?.get(definition)?.jsonObject
            ?: fail("Definition '$definition' not found in $schemaFile")
    }

    fun assertConformsTo(schema: JsonObject, instance: JsonElement) {
        val violations = validate(schema, instance, "$")
        if (violations.isNotEmpty()) {
            fail("Payload violates contract:\n  ${violations.joinToString("\n  ")}\nPayload: $instance")
        }
    }

    fun validate(schema: JsonObject, instance: JsonElement, path: String): List<String> {
        val violations = mutableListOf<String>()

        schema["const"]?.let { if (it != instance) violations += "$path: expected const $it, got $instance" }
        schema["enum"]?.jsonArray?.let { if (instance !in it) violations += "$path: $instance not in enum $it" }

        schema["type"]?.let { typeNode ->
            val allowed = when (typeNode) {
                is JsonArray -> typeNode.map { it.jsonPrimitive.content }
                else -> listOf(typeNode.jsonPrimitive.content)
            }
            if (allowed.none { matchesType(it, instance) }) {
                violations += "$path: expected type $allowed, got ${describe(instance)}"
            }
        }

        if (instance is JsonPrimitive && instance.isString) {
            when (schema["format"]?.jsonPrimitive?.contentOrNull) {
                "uuid" -> runCatching { UUID.fromString(instance.content) }
                    .onFailure { violations += "$path: '${instance.content}' is not a uuid" }
                "date-time" -> runCatching { OffsetDateTime.parse(instance.content) }
                    .onFailure { violations += "$path: '${instance.content}' is not an RFC 3339 date-time" }
            }
        }

        if (instance is JsonObject) {
            val properties = schema["properties"]?.jsonObject ?: JsonObject(emptyMap())
            schema["required"]?.jsonArray?.forEach { req ->
                val name = req.jsonPrimitive.content
                if (name !in instance) violations += "$path: missing required property '$name'"
            }
            instance.forEach { (name, value) ->
                val propertySchema = properties[name]?.jsonObject
                when {
                    propertySchema != null -> violations += validate(propertySchema, value, "$path.$name")
                    schema["additionalProperties"] is JsonObject ->
                        violations += validate(schema["additionalProperties"]!!.jsonObject, value, "$path.$name")
                    schema["additionalProperties"]?.jsonPrimitive?.booleanOrNull == false ->
                        violations += "$path: property '$name' is not declared in the contract"
                }
            }
        }

        if (instance is JsonArray) {
            schema["minItems"]?.jsonPrimitive?.intOrNull?.let { min ->
                if (instance.size < min) violations += "$path: expected at least $min items, got ${instance.size}"
            }
            schema["items"]?.jsonObject?.let { itemSchema ->
                instance.forEachIndexed { i, item -> violations += validate(itemSchema, item, "$path[$i]") }
            }
        }

        return violations
    }

    private fun matchesType(type: String, instance: JsonElement): Boolean = when (type) {
        "object" -> instance is JsonObject
        "array" -> instance is JsonArray
        "null" -> instance is JsonNull
        "string" -> instance is JsonPrimitive && instance.isString
        "boolean" -> instance is JsonPrimitive && !instance.isString && instance.booleanOrNull != null
        "integer" -> instance is JsonPrimitive && !instance.isString && instance.content.toLongOrNull() != null
        "number" -> instance is JsonPrimitive && !instance.isString && instance.content.toDoubleOrNull() != null
        else -> false
    }

    private fun describe(instance: JsonElement): String = when (instance) {
        is JsonObject -> "object"
        is JsonArray -> "array"
        is JsonNull -> "null"
        is JsonPrimitive -> if (instance.isString) "string" else "literal ${instance.content}"
    }

    private fun sharedSchemaPath(schemaFile: String): Path {
        var dir: Path? = Paths.get("").toAbsolutePath()
        while (dir != null) {
            val candidate = dir.resolve("shared/events/schemas/$schemaFile")
            if (Files.exists(candidate)) return candidate
            dir = dir.parent
        }
        fail("Could not locate shared/events/schemas/$schemaFile above ${Paths.get("").toAbsolutePath()}")
    }
}
